### ISSUE TITLE
When Dropout layer is shared Siamese-style, dropped units are not synchronized.

### ISSUE THREAD

**Issue by ozabluda** [keras-team/keras#8802]
When sharing `Dropout` layers Siamese-style, I wasn't able to synchronize dropped units. For example, in the code below, `noise_shape` has no effect. `seed` parameter has no effect either. Shared `Dropout` layers should synchronize dropped units by default, otherwise they are not shared in any meaningful way.
```
# https://gist.github.com/ozabluda/bbc6c84c0e69bfd9ca55170fd3ab040d
# https://github.com/keras-team/keras/issues/8802

from keras.models import Sequential, Model
from keras.layers import Dense, Dropout, Input, Lambda
from keras import backend as K
import numpy as np

m = Sequential([
    Dropout(rate=0.5, input_shape=(1,), noise_shape=(1,1))
])
m.summary()

input_a = Input(shape=(1,))
input_b = Input(shape=(1,))

processed_a = m(input_a)
processed_b = m(input_b)

def l1_distance((x1, x2)):
    return K.sum(K.abs(x1-x2), axis=1)

c = Lambda(l1_distance, output_shape=(1,))([processed_a, processed_b])
s = Model([input_a, input_b], c)
s.compile(optimizer='sgd', loss='mse')
s.summary()

x0 = np.array([1])
x1 = np.array([1])
x  = [x0,x1]
y  = np.array([0])

s.fit(x, y, verbose=1, epochs=10)

print(s.evaluate(x,y), s.predict(x))
```
Output:
```
Epoch 1/10
1/1 [==============================] - 1s 1s/step - loss: 0.0000e+00
Epoch 2/10
1/1 [==============================] - 0s 3ms/step - loss: 4.0000
Epoch 3/10
1/1 [==============================] - 0s 3ms/step - loss: 4.0000
Epoch 4/10
1/1 [==============================] - 0s 4ms/step - loss: 0.0000e+00
Epoch 5/10
1/1 [==============================] - 0s 3ms/step - loss: 0.0000e+00
Epoch 6/10
1/1 [==============================] - 0s 3ms/step - loss: 0.0000e+00
Epoch 7/10
1/1 [==============================] - 0s 3ms/step - loss: 4.0000
Epoch 8/10
1/1 [==============================] - 0s 3ms/step - loss: 4.0000
Epoch 9/10
1/1 [==============================] - 0s 3ms/step - loss: 4.0000
Epoch 10/10
1/1 [==============================] - 0s 4ms/step - loss: 0.0000e+00
1/1 [==============================] - 0s 14ms/step
(0.0, array([ 0.], dtype=float32))
```

Note that
1. during inference nothing is dropped, so loss is 0, as expected
2. if dropout rate=0, the loss is 0 during training, as expected
3. Note that on Epochs 2,3,7,8,9 the loss is 4, which I don't understand at all. Maybe another bug, I need to investigate further.

**Comment by fchollet** [comment]
Dropout is a random operation, so the dropout pattern is unique for every data flow.

To create a dropout pattern that will be shared across different data flows, you can create a custom layer, which would do the following:

- maintain a non-trainable weight `dropout_mask`, the size of the layer input, initialized to ones
- multiples its inputs by `dropout_mask` (times a scaling factor) in train mode only, in `call`
- adds an update op to vary `dropout_mask` at every step. You can also make this update be train-mode-only.

Something like;

```python

class MyDP(Layer):
    def __init__(self, rate, **kwargs):
       self.rate = rate
       super(...).__init__(**kwargs)

    def build(self, input_shape):
       self.dropout_mask = self.add_weight(shape=input_shape, initializer='ones', trainable=False)

   def call(self, inputs):
      dropped =  K.in_train_phase(self.dropout_mask * inputs * ...)
      if not self.updates:
         self.add_update(self.dropout_mask, K.dropout(self.dropout_mask * 0 + 1, self.rate)
     return droppped
```

Warning: untested, incomplete, etc

**Comment by ozabluda** [comment]
I think that item 3 above (loss=4) is due to the fact that Dropout multiplies activations by `1-p` during training, a well-known trick to avoid dividing during inference). So that's correct.

With unsynchronized masks, shared Dropout layers are not actually shared in any meaningful way. I can't tell if it's a bug/misfeature in Keras or the TF backend (is so, I'll file the bug report there). 

This was unexpected to me, and I came to the realization that masks may not be shared accidentally, while debugging Siamese network training. When fed identical data, two networks with supposedly shared everything, produce different final activations.

Apparently, I have multiple sources of stochasticity, which I resolve one-by-one. The ones I identified so far are `Dropout` and `BatchNormalization`, but there could be more.

Replacing standard dropout layer with a custom one is awkward, because the base network is pretrained, so I'd need to do network surgery, which I am not entirely sure how to.

BTW, setting `seed` does appear to work after all (there was a bug in my earlier version of the test).

**Comment by ozabluda** [comment]
`seed` also survives `save_model()/load_model()`, so that gives a workable workaround.

**Comment by lschaupp** [comment]
@ozabluda quick question regarding the seed: does the dropout layer for each training epoch drop (once the seed is given) drop the same neurons? Or how is the seed to interpret in this case?

**Comment by ozabluda** [comment]
almost certainly not, since it would make no sense and would be a severe bug, not likely to go unnoticed.

**Comment by hoangcuong2011** [comment]
I think the issue @ozabluda posted is very good/important. If we have a Siamese network, there is a possibility that it is risky to apply certain techniques like Dropout and Batch Normalization. I personally tried this in a Siamese network and I noticed the performance was very bad!  

Do you have any idea how to improve this yet @ozabluda? Thanks!

**Comment by hoangcuong2011** [comment]
OK so I implemented a shared Dropout layer by myself, similar to what @fchollet suggested.

`
class SharedDropout(Layer):
	# learnt this from this link: https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/ops/nn_ops.py
	def __init__(self, keep_prob_rate=0.5, **kwargs):
		self.keep_prob_rate = keep_prob_rate
		super(SharedDropout, self).__init__(**kwargs)

	def build(self, input_shape):
		super(SharedDropout, self).build(input_shape)

	def call(self, inputs):
		input_left = inputs[0]
		input_right = inputs[1]
		random_tensor = self.keep_prob_rate
		random_tensor += tf.random_uniform(tf.shape(input_left), dtype=input_left.dtype)
		binary_tensor = tf.floor(random_tensor)

		def DropoutLeft():
			ret_left = tf.divide(input_left, self.keep_prob_rate) * binary_tensor
			return ret_left

		def DropoutRight():
			ret_right = tf.divide(input_right, self.keep_prob_rate) * binary_tensor
			return ret_right

		return [K.in_train_phase(DropoutLeft, input_left, training=None), K.in_train_phase(DropoutRight, input_right, training=None)]

	def compute_output_shape(self, input_shapes):
		return [(input_shapes[0][0], input_shapes[0][1]), (input_shapes[1][0], input_shapes[1][1])]

`
Example to call the layer:
`x_input_left = Input(shape=(10,), name='x_input_left')
	x_input_right = Input(shape=(10,), name='x_input_right')
	shared_dropout = SharedDropout()
	x_left, x_right = shared_dropout([x_input_left, x_input_right])`

It took me a few hours to implement/test this. So I am proud of this for a moment :D
